### PR TITLE
matcher: Generic matching API

### PR DESCRIPTION
**Pull request by tyxia**
**Background:** 
The current matching API is limited to a single string due to the requirement for both the  return of [data input result](https://github.com/envoyproxy/envoy/blob/fa5f548d05da6495c175b5c124292fa0bae215de/envoy/matcher/matcher.h#L206) and the input to [input matcher](https://github.com/envoyproxy/envoy/blob/fa5f548d05da6495c175b5c124292fa0bae215de/envoy/matcher/matcher.h#L164) to be single string.  A more flexible and generic matching API is desired so that we can perform matching on various data types. In general, we want to support opaque data that is specific to the matcher.  For example, we want to have CEL matcher where CEL expression is plugged into the unified matching API.

**Changes:**
This PR implements the templatization of the data input, making it the first step in supporting the following configuration while minimizing the impact/changes on existing code.
```
matcher_list:
    matchers:
      predicate:
        single_predicate:
          input:
            typed_config:
              "@type": type.googleapis.com/xds.type.matcher.v3.HttpAttributesCelMatchInput
          custom_match:
             typed_config: 
              "@type" : type.googleapis.com/xds.type.matcher.v3.CelMatcher
               expr_match:
                  checked_expr: ……
```

Several PRs will follow: 1)Templatizing. the input matcher  2) Utilizing the generic unified match API with a CEL matcher/expression (a proof of concept has been tested locally and confirmed to be working). After these changes, it's possible that further adjustments will be needed to make the entire matching API generic. However, the solution implemented here is highly extensible. Any additional changes required will be addressed in separate PRs.

Note: string default template argument is used because it is most common type in codebase. This choice also minimizes the impact on existing code, as it requires fewer code changes. Additionally, in cases where no use cases for other data types are present, or when no use cases are needed at the moment, we default to using the "std::string" data type.

Risk Level: low, no functional change to existing code
Testing: CD/CI

Signed-off-by: tyxia [tyxia@google.com](mailto:tyxia@google.com)

### CODE REVIEW

**Comment by tyxia** [comment]
/assign @snowp, @kyessenov  
as the owner of matcher

File: envoy/matcher/matcher.h [hunk 1]
```diff
@@ -187,63 +196,67 @@ class InputMatcherFactory : public Config::TypedFactory {
...
+    case DataAvailability::AllDataAvailable:;
     }
     return out;
   }
 };
 
+// Template deduction guide for DataInputGetResult.
+template <class ResultDataType>
+DataInputGetResult(DataAvailability data_availability, ResultDataType data)
+    -> DataInputGetResult<std::string>;
```
Comment (kyessenov) [inline envoy/matcher/matcher.h:232]: I'm not sure I understand this guide - it sounds like it's saying that any constructed `DataInputGetResult` is actually `DataInputGetResult<string>`?

Comment (tyxia) [reply envoy/matcher/matcher.h:232]: This guide to help compiler in certain cases to deduce the template arguments without specifying `std::string` , For instance, we don't need to have `DataInputGetResult<std::string>` in the createMatchers field_matcher_test.cc.

If we want to use other data type, we just need to specify `ResultDataType`. The use case for such will be added in my following PRs when add the support for CEL matcher with this matching API.

A quick draft example in same field_matcher_test.cc for illustrative purpose, in addition to existing string `createMatchers`, we can have 
```
std::vector<FieldMatcherPtr<TestIntData>>
  createIntMatchers(std::vector<std::pair<bool, DataAvailability>> values) {
    std::vector<FieldMatcherPtr<TestIntData>> matchers;
    matchers.reserve(values.size());
    for (const auto& v : values) {
      matchers.emplace_back(std::make_unique<SingleFieldMatcher<TestIntData, int>>(
          std::make_unique<TestIntInput>(DataInputGetResult<int>{v.second, absl::nullopt}),
          std::make_unique<BoolMatcher>(v.first)));
    }
    return matchers;
  }
```

File: envoy/matcher/matcher.h [hunk 2]
```diff
@@ -187,63 +196,67 @@ class InputMatcherFactory : public Config::TypedFactory {
...
       break;
-    case DataInputGetResult::DataAvailability::AllDataAvailable:;
+    case DataAvailability::AllDataAvailable:;
     }
     return out;
   }
 };
 
+// Template deduction guide for DataInputGetResult.
+template <class ResultDataType>
```
Comment (htuch) [inline envoy/matcher/matcher.h:230]: The other alternative here is we make the input type some oneof union. It could be a string as today, or it could be a protobuf dynamic type, which supports a fairly wide variety of data models. The main downside there is performance, which we probably should be sensitive to on the match path, so I think I'm convinced that doing this at C++ level makes sense. Not certain though this needs to be templates, will defer to @snowp.

Comment (kyessenov) [reply envoy/matcher/matcher.h:230]: In my last attempt, I explored `std::any` and `std::variant`. The first one is a terrible choice, since it's hard to work with. Variant could work but it makes the input structure wide and not open-ended. Ultimately, the main difference is in the way matchers adapt to inputs of certain types. It can be done with a big switch on `input.type` or via template substitution `template <Type> match(DataInputGetResult<Type>)`. If we go with templates, I want to make sure we have a clear way to support integers - I think template specialization for each numeric type is too error-prone.

File: source/extensions/matching/network/common/inputs.h [hunk 3]
```diff
@@ -66,13 +66,12 @@ DECLARE_FACTORY(HttpDestinationIPInputFactory);
...
-  Matcher::DataInputGetResult get(const MatchingDataType& data) const override {
+  Matcher::DataInputGetResult<std::string> get(const MatchingDataType& data) const override {
     const auto& address = data.localAddress();
     if (address.type() != Network::Address::Type::Ip) {
-      return {Matcher::DataInputGetResult::DataAvailability::AllDataAvailable, absl::nullopt};
+      return {Matcher::DataAvailability::AllDataAvailable, absl::nullopt};
     }
-    return {Matcher::DataInputGetResult::DataAvailability::AllDataAvailable,
-            absl::StrCat(address.ip()->port())};
+    return {Matcher::DataAvailability::AllDataAvailable, absl::StrCat(address.ip()->port())};
```
Comment (kyessenov) [inline source/extensions/matching/network/common/inputs.h:74]: This should really be an integer input value, not a string. Can we try changing it and see how matchers adapt to it?

File: envoy/matcher/matcher.h (hunk 1 above)
Comment (kyessenov) [reply envoy/matcher/matcher.h:232]: Yeah, I think I see it as a default compiler hint - meaning it'll try to deduce string as soon as possible, and that means any other type value but the string requires a copy of code - `createIntMatchers` instead of generic `createMatchers`. I don't have a clear grasp of how much code needs to be explicitly instantiated since all inputs are strings -- hence, my concern.

Comment (tyxia) [reply envoy/matcher/matcher.h:232]: `createIntMatchers` here is the illustrative example (and in test code). I think If it is set to be int type, there will be no need of the impl for the other type.

From my prototype for CEL with all the changes requires, there will not have copy of code in the non-test code path as they are all handled in place with templatized  argument and we just need to specify the template argument when we use it. I will share more details in the doc i mentioned below.

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (tyxia) [reply envoy/matcher/matcher.h:230]: Thank you both for the review and inputs!

Re htuch: at C++ level,  I think template is also the one aligned with existing matching API, similar to how `DataType` template is used.

Re kyessenov: I mostly focused on supporting CEL though I did test with INT in my prototype.

I am definitely open to other ideas. It is probably better to start with the design doc which includes whole picture and more details(e.g., use example to both data input and input matcher) of this approach. I will share with you later for discussion and review.

File: source/extensions/matching/network/common/inputs.h (hunk 3 above)
Comment (tyxia) [reply source/extensions/matching/network/common/inputs.h:74]: Yes, but as i mentioned in the PR description, I plan to do it in following PRs as this essentially requires second part of changes --- make input matcher interface generic. 

So the reason I don't change it here now is 1) to avoid making the PR too large to review and 2) to fit the timeline of two import projects that depends on this API with CEL. I plan to do INT case after CEL use case as CEL should be able provide good idea of how this API is being used.

I fully agree that it is better to give the reviewers whole picture and more details , so I will write up a doc for review and discussion.

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (snowp) [reply envoy/matcher/matcher.h:230]: Originally we went with templates because it worked nicely with the way factory registrations work: `RequestHeaderInput` would sub class `DataInput<HttpData>` and be a distinct registration that wouldn't work with something looking for `DataInput<TcpData>` registrations. One of the complications from this though was that we ended up with a lot of repeated registrations whenever we wanted to register the factories for different kind of inputs. 

So learning from that, it would be nice that we try to avoid making extension registration/implementation even more complicated as we're starting to support multiple output types. 

It might be that we find that templates is the only efficient way of doing this (std::variant has some overhead), but I'd be interested to see what it would look like with std::variant. I think you could get pretty far with some helper functions/base classes that would hide the std::variant unpacking from the extension writer

tl;dr a doc would be appreciated :)

File: envoy/matcher/matcher.h (hunk 1 above)
Comment (kyessenov) [reply envoy/matcher/matcher.h:232]: Ack, thanks for confirming.

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (kyessenov) [reply envoy/matcher/matcher.h:230]: From the point of view of consumers of inputs, they are mostly interested in a projection to the desired type, e.g. to string or to int or to CEL environment. So we can use regular interfaces to express this without templates I think. I've looked into how CEL represents values in C++ - there's a lot of useful design ideas there.

Another important property we probably want - either make Input non-copyable or make it trivially copyable by moving the storage elsewhere. A large object like CEL environment is undesirable to copy.

**Review by htuch — COMMENTED** [review]
The basic idea here LGTM, but I'll defer to @kyessenov and @snowp on ensuring we have the right C++ typing construct.

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (tyxia) [reply envoy/matcher/matcher.h:230]: I think runtime check (like a big switch on input.type mentioned above) is not good for performance.  

And yes, I did change the interface of input matcher to take in `const ref` instead of `absl::optional` exactly to avoid copy of large object like cel environment.(will be shown in next PR though as this PR only has change for dataInput) 

I have shared the design doc with you all. Please take a look. Thank you!

**Comment by kyessenov** [comment]
Sorry about the delay - I'm reviewing the doc. My main goal is to ensure that it's both easy to read/write and not too restrictive. I think the templates method forces the client to know which type to use - that means RateLimit can only use CEL and other matcher can only use non-CEL. It's be better if the existing matchers would allow CEL with little code change.

File: envoy/matcher/matcher.h [hunk 4]
```diff
@@ -187,63 +196,67 @@ class InputMatcherFactory : public Config::TypedFactory {
...
-
+template <class ResultDataType = std::string> struct DataInputGetResult {
   DataAvailability data_availability_;
   // The resulting data. This will be absl::nullopt if we don't have sufficient data available (as
   // per data_availability_) or because no value was extracted. For example, consider a DataInput
   // which attempts to look a key up in the map: if we don't have access to the map yet, we return
   // absl::nullopt with NotAvailable. If we have the entire map, but the key doesn't exist in the
   // map, we return absl::nullopt with AllDataAvailable.
-  absl::optional<std::string> data_;
+  absl::optional<ResultDataType> data_;
```
Comment (yanavlasov) [inline envoy/matcher/matcher.h:206]: How are we going to ensure that the type produced by the `input` extension is suitable for the `custom_match` extension. Also the `exact_match_map` and `prefix_match_map` expect `input` to produce a string. What happens if the input type is something other than string?

**Comment by tyxia** [comment]
This PR is closed in favor of the approach described in this design doc: [Envoy Generic Matching API](https://docs.google.com/document/d/1Ll_ylq90Z0D2vZOhWHXQksK82mN2bstfV5o4goCigTs/edit?resourcekey=0-63lyb1jUu_68P6BG7rqlAg#heading=h.xgjl2srtytjt)

File: envoy/matcher/matcher.h (hunk 2 above)
Comment (htuch) [reply envoy/matcher/matcher.h:230]: Putting aside performance for a second, I'm wondering why we can't just use a solution similar to `FilterState` and at config validation time some RTTI to validate compatibility of input-matcher types? Then maybe variant is interesting as some kind of optimization for unboxed types but not inherently needed?

Looking at the code, I'm definitely in agreement with @snowp that we don't want to increase the level of template type complexity if we can avoid it, it's already quite hard to follow the types and instance-template relationships when paging back in after some time.
